[NPISO] SFC/SNES multitap optimization

DigiDwrf opened the pull request:
Hey Darthcloud, I finally optimized this multitap code, no blinking on the 4th and the 5th controllers anymore. Every game that was glitchy now works fine. Latch code was fixed too, and top gear 3000 now works (some B button is pressed in the menu at times randomly, but it happens not very often), no blink anymore and all 4 players are detected. I've tried it on Bomberman games, Pieces, Micromachines and EA sports games, they all work fine. I've been working on this code, day and night, I've reprogrammed my ESP32 more than 300 times (no lie), and finally nailed it. The only thing I can't get rid off is a message that sometimes appears at the beginning on Pieces that says that the mouse shouldn't be connected into the multitap (don't know why, since the 15th cycle is always high, this info comes from the controller!), but once I reset the esp32 (or replug the port 2) the game loads normally, it doesn't appear never again, and all 5 controllers work perfect with no blink on the 4th and the 5th controllers.

I think that part of the problem was the time the ESP32 takes (cycles, I guess), so I made a "big" `switch (idx[1])` package, using the `while` command before `set_data` void  in all cases to take advantage of the free time after `low_io & P2_CLK_MASK` or `high_io & NPISO_LATCH_MASK` conditions.

I'm really happy :) This project is awesome, you rock!

darthcloud commented:
Will merge once I test it and review, might take couple weeks.

Thanks for looking into this, I pretty much gave up on it.

DigiDwrf commented:
Checked some last stuff. Now it's ready, I hope.

I'm gonna branch a 8p support, still working on N-Warp Daisakusen. It may never work, since the timing is VERY tight, same for Super Formation Soccer games: 2p moves 4p, and 3p shoulder buttons go crazy, so there's no much to do here.

At this point, every other game should work fine (if clock is not too fast), tested a lot now and all they work OK. EA latch glitch is not a problem anymore because P2-D1 is always low when latch is high.

DigiDwrf commented:
Still working on some test, I'll let you know when it's ready

darthcloud commented:
Tested this morning with top gear 3000 and I gets randoms output from P2-P4

How are you setup for testing this? It's wired with 74AHCT125 level shifter like this ?
https://github.com/darthcloud/BlueRetroHW/blob/master/DIY/SNES.pdf

darthcloud commented:
Super Bomberman 2 & NHL 94 look fine however, but Top Gear 3000 is unplayable.

DigiDwrf commented:
> Super Bomberman 2 & NHL 94 look fine however, but Top Gear 3000 is unplayable.

Sure, still working on some tests by now, will be have it ready soon.

DigiDwrf commented:
> Tested this morning with top gear 3000 and I gets randoms output from P2-P4
> 
> How are you setup for testing this? It's wired with 74AHCT125 level shifter like this ? https://github.com/darthcloud/BlueRetroHW/blob/master/DIY/SNES.pdf

btw, all my wiring corresponds to your data sheets.

DigiDwrf commented:
@darthcloud There's something I'd like to implement: Since SNES games detect controller presence on plug, I'd like to `set_data(x, x, 1)` on 17th clock (for NO controller) when there's no bluetooth device connected on a particular port, and change that value to 0 (plugged controller) when I pair a new controller. was looking for that particular variable but had no luck (e. g. `bt_adapter.data[x].etc...` ) can you gimme a hand with that? it would be awesome to virtually plug new paired controllers into the snes, know what I mean?

darthcloud commented:
Any reason for this?, I rather keep them always on personally.

The proper way would be set the SNES as hotplug in the system manager
https://github.com/darthcloud/BlueRetro/blob/c83448b2e62ce1a83cd643f0743aecfa3f24c9ed/main/system/manager.c#L598

and define a snes_port_cfg function to handle enabling/disabling the port:
https://github.com/darthcloud/BlueRetro/blob/c83448b2e62ce1a83cd643f0743aecfa3f24c9ed/main/wired/wired_comm.c#L87

But unless there a good reason again I rather keep port always on. Some games may look for controllers only at boot time.

DigiDwrf commented:
I understand. I don't know or have seen any game that look for controllers only at boot yet, some of them look for the multitap or controllers only before entering the 1p/2p/3p/4p/... screen. I've seen this "on boot" detection in a N64. But, if that's really the case, it should be left on then.
The reason for this is that it would be easier, I think, to see how many controllers are connected and propperly working on the game screen, so, the multitap would be ON from the start, but each controller could be virtually "plugged" into it when a new device is added during gameplay. Thanks for the info 👍

DigiDwrf commented:
@darthcloud hey! Now it's REALLY ready for testing.

I finally implemented a 2p / 5p autoswitch for some 2p games (like Battletoads in Battlemaniacs) that sends a low SELECT signal, and messes up with the code. No reason for developers to do this in 2p games, but yet, it's a problem, a glitch I guess. I think that's why the multitap has a physical 2p/5p switch on it too, not only to avoid games that shows any warning message, but also to avoid any failure. Without this switch, the 2nd controller works with the 4th paired one in this games (of course, you can "unplug" the multitap inside the blueretro.io menu, but it's kinda tedious if you want to switch fast between games, so that's the reason I made this switch).

The switch works like this: It starts as 5p mode and starts counting until it reaches cero, and enters 2p mode. If there's more clocking sent by software on the 2nd port (from 18th clock and up) if switches back to 5p mode. This switch won't work for warning messages that appears on boot inside 2p games like Yoshi's Island because this games look for the multitap, and this switch will turn on when it's requested by the console. The 2nd controller plug can be replugged after this message and it won't bother again (this depends on the game, I guess).

intermittent detecton on the 4th and 5th controllers is not present anymore also, not found any game that I've tested with this problem anymore.

Top Gear 3000 finally works!!!!

This code is heavily optimized so I think it should work OK for most games, but It's impossible to make it work propperly in very few games that latch and clock beyond this device's GPIO speed.

DigiDwrf commented:
Hey @darthcloud, after some heavy testing, I realized that this 2p switch doesn't work in some EA games (they look for multitap presence when clock goes low and latch is high before looking for controllers). When I implemented this clock detection, the switch stopped working, and stuck always in 5p mode inside 2p games with 5p signaling on them.

So, I've changed it by a manual switch: just have to press and hold B, DOWN, L, and R between 1 and 2 seconds on the 2nd wireless controller to activate 2p mode, and pressing B the 3rd wireless controller will switch it back to 5p mode. 5p mode comes activated by default.

In this way, 2p games that presents 5p signaling (3rd port on multitap acts like the 2nd port on the console) and games that are 2p but intentionally look for the multitap at boot, asking to disconnect any device different than controllers, can be bypassed with no need to change the setup inside the [blueretro.io](url) panel.

This is very useful for fast game swapping when playing on a multicart like everdrive or sd2snes.

So now, this is ready, you can test this and let me know what do you think about it.

DigiDwrf commented:
@darthcloud hey, could you test this last revision (bf67916)?  I'm sorry that you had to test a work in progress before, I was still fixing some stuff and didn't let you know, my bad! Btw, some very few games will never work, but intermittent detection of multitap/controllers isn't present anymore in any game I've tested so far, so it's an improvement anyway.

darthcloud commented:
No worries, I'm busy finishing v1.8 release, I'll get to this PR after.

Just leave as-is, I'll take care of final integration.

darthcloud commented:
I FU when I pushed my change into your branch and it closed this PR.

Can you review this commit:
https://github.com/darthcloud/BlueRetro/commit/10a47b1c69e82694d7b0581b2740824d92ad62d3

You had a while loop that always returned on the first iteration so I changed it for a if.

I retested regular NES/SNES to make sure no regression was present.

I tested multitap with TopGear, Bomberman 2 and NHL 94. They all looked fine.

I will merge change once you confirm my change look fine.

Thanks! sorry for long delay on this!

DigiDwrf commented:
awesome! I'm gonna check it out.

DigiDwrf commented:
> I FU when I pushed my change into your branch and it closed this PR.
> 
> Can you review this commit:
> https://github.com/darthcloud/BlueRetro/commit/10a47b1c69e82694d7b0581b2740824d92ad62d3
> 
> You had a while loop that always returned on the first iteration so I changed it for a if.
> 
> I retested regular NES/SNES to make sure no regression was present.
> 
> I tested multitap with TopGear, Bomberman 2 and NHL 94. They all looked fine.
> 
> I will merge change once you confirm my change look fine.
> 
> Thanks! sorry for long delay on this!

Ok, I've reviewed it. The reason I put a while loop there was to stick on that part of the code until latch goes high again, so it could save some cycles and set data in time, as much as possible. I understood that the while loop would repeat that part of code, until that particular condition becomes false. I'm not an expert, so I'm not quite sure if the return command breaks the loop and makes the code to start from the begining of the void, so in that case, It'll would be unnecessary to use while. I tried it at first using the if condition, and there was some blinking detection issues, and then, when I tried with the while loop, it was a major fix!!! it's practically the "heart" of that part of code. Have you tried this code with an if instead of the while loop already? Let me know! I'll give it a try again, I been working and tweaking this code for months, so I am quite sure that this while loop is necessary.

darthcloud commented:
I tried my version only and it look fine.

With the unconditional return within the while its impossible it ever loop.

the next interrupt will start beginning function.

DigiDwrf commented:
@darthcloud hey! just tested the `if` condition code and it works 👍

DigiDwrf commented:
@darthcloud So, this is finally ready. There's something I've been thinking, do you think that the 5p mode should be bypassed with the second controller, or the first one (or both)? For example, I put on Yoshi's Island and the multitap is set on slot 2 inside the blueretro.io config menu. The "disconnect device" message appears. Then I have to connect a second controller to turn off 5p mode and then the game runs normally. Maybe the switch should be controlled with the fist controller only (or maybe any controller). What do you think about it? Btw, happy new year!

darthcloud commented:
I think its a non issue. I my mind switching mode require power cycle.

darthcloud commented:
I merged the PR this morning.

DigiDwrf commented:
Awesome! thanks for letting me contribute with your project :)
